### DISCUSSION TITLE
Add custom props and use it into presets

### DISCUSSION THREAD

**Discussion by AnthonyRuelle** [primefaces/community#2433] [PrimeVue] [answered]
Hi, 

How do I add a custom new prop to a Prime component and use it in tailwind presets file ?

I need to do this on several components and cannot access my personal props in the presets. 

Thanks

**Comment by AnthonyRuelle** [comment]
Up please, any ideas ??

**Reply by j0rgedev** [reply]
Hello @AnthonyRuelle. Could you provide an example? So, we could understand more what your needs are.

**Reply by AnthonyRuelle** [reply]
Hello and thank you for your answer.

I have lots of components where I would like to do this. For example, on the Tag component I would like to pass a custom prop called range which would allow me to associate a particular style with these tags according to the values ​​of this prop. 
Props passed do not allow me to have the freedom I need

**Reply by j0rgedev** [reply]
Oh I see. What you can do is create a wrapper of the component. For example, I create MyTag Component which will be a wrapper for PrimeVue Tag Component. It'll look something like this:

**MyTag.vue**
```vue
<script setup>
import Tag from 'primevue/tag'

const props = defineProps({
  range: Number,
})

const severity = computed(() => {
  if (props.range < 20) {
    return 'danger'
  }
  else if (props.range < 60) {
    return 'warn'
  }
  else {
    return 'success'
  }
})
</script>

<template>
  <Tag
    v-bind="$attrs"
    :severity="severity"
  />
</template>
```

**main.vue**
```vue
<template>
  <MyTag
    range="100"
    value="Test Tag"
    <!-- More PrimeVue Tag props here -->
  />
</template>
```
With [$attrs](https://vuejs.org/guide/components/attrs), MyTag can accept any attribute and pass them on to the PrimeVue component, saving you from defining every single prop.
I've also seen that all PrimeVue components have an attrs name inside the pt attribute; however, I'm not familiar with it.

**Reply by AnthonyRuelle** [reply]
Hi :)

This does not solve my problems because I need to have new severity values ​​for example to manage my different cases. 
Either new property values ​​or new props which can be directly used in the preset files but I can't find anything on this subject.
In your example you define the following severity my props but I want them to be separate values

**Reply by j0rgedev** [reply]
Oh, I got it. So, if you mean like extending the severity prop to add yours and then styling it in the preset files, I guess that's not supported yet. So sorry for the misunderstanding

**Reply by AnthonyRuelle** [reply]
Yes indeed.. I have hope that there is a possibility but that it is just not documented

**Reply by rogrile** [reply]
Hey, I know this is for sure supported, but it's not going to be called props, but something like attributes. I think I've seen this on the discord server and I made use of it at some point. Let me check and I'll come back

**Reply by georgebrianb** [reply, marked as answer]
You can pass an attribute and refer to it like this:

```js
// app.vue

<template>
  <ThemeSwitcher />
  <div class="card flex justify-center">
    <InputText v-model="value" :my-invalid-prop="true" />
  </div>
</template>

<script setup>
import { ref } from 'vue';

const value = ref(null);
</script>




// presets/lara/InputText/index.js  - or whatever path you have for your Preset/PassThrough

export default {
  root: ({ props, context, attrs }) => ({
    class: [
      {
        'border-black': attrs['my-invalid-prop'], // <<<< ---------  here you go
      },
      'font-sans leading-none',
      'm-0',
      {
        'px-4 py-4': props.size == 'large',
        'px-2 py-2': props.size == 'small',
        'p-3': props.size == null,
      },
      'text-surface-600 dark:text-surface-200',
      'placeholder:text-surface-400 dark:placeholder:text-surface-500',
      'bg-surface-0 dark:bg-surface-900',
      'border border-surface-300 dark:border-surface-600',
      {
        'hover:border-primary-500 dark:hover:border-primary-400':
          !context.disabled,
        'focus:outline-none focus:outline-offset-0 focus:ring focus:ring-primary-500/50 dark:focus:ring-primary-400/50':
          !context.disabled,
        'opacity-60 select-none pointer-events-none cursor-default':
          context.disabled,
      },
      'rounded-md',
      'appearance-none',
      'transition-colors duration-200',
    ],
  }),
};
```

**Reply by j0rgedev** [reply]
@georgebrianb exactly! That's what I meant when I said there was an attrs param but I didn't know how to use it. Thank you. Haven't tested it but it seems it will work.

**Reply by AnthonyRuelle** [reply]
> You can pass an attribute and refer to it like this:
> 
> ```js
> // app.vue
> 
> <template>
>   <ThemeSwitcher />
>   <div class="card flex justify-center">
>     <InputText v-model="value" :my-invalid-prop="true" />
>   </div>
> </template>
> 
> <script setup>
> import { ref } from 'vue';
> 
> const value = ref(null);
> </script>
> 
> 
> 
> 
> // presets/lara/InputText/index.js  - or whatever path you have for your Preset/PassThrough
> 
> export default {
>   root: ({ props, context, attrs }) => ({
>     class: [
>       {
>         'border-black': attrs['my-invalid-prop'], // <<<< ---------  here you go
>       },
>       'font-sans leading-none',
>       'm-0',
>       {
>         'px-4 py-4': props.size == 'large',
>         'px-2 py-2': props.size == 'small',
>         'p-3': props.size == null,
>       },
>       'text-surface-600 dark:text-surface-200',
>       'placeholder:text-surface-400 dark:placeholder:text-surface-500',
>       'bg-surface-0 dark:bg-surface-900',
>       'border border-surface-300 dark:border-surface-600',
>       {
>         'hover:border-primary-500 dark:hover:border-primary-400':
>           !context.disabled,
>         'focus:outline-none focus:outline-offset-0 focus:ring focus:ring-primary-500/50 dark:focus:ring-primary-400/50':
>           !context.disabled,
>         'opacity-60 select-none pointer-events-none cursor-default':
>           context.disabled,
>       },
>       'rounded-md',
>       'appearance-none',
>       'transition-colors duration-200',
>     ],
>   }),
> };
> ```

Wow, perfect, it's working fine 

Maybe there should be this kind of explanation in the doc because it's super interesting and useful..

**Reply by timeax** [reply]
Is there a way to prevent custom props (attributes) from being added to the browser DOM? in the browser DOM?
